### PR TITLE
feat: Split `lerna version` from of `lerna publish`

### LINKED ISSUE #277: [Feature] Canary should have incrementing version numbers.
The canary feature is fantastic, however because it uses commit hashes instead of numbers, publishing a canary release with the same basic version number might require the wrong versions when using ^ for all the other package dependencies. Can we either overwrite the dependencies to require explicit canary versions OR properly version canaries so that newer canaries will always be installed over older ones? :)

### LINKED ISSUE #936: Lerna Publish pre-patch/minor/major behavior
First I just want to say this is a great project! Thanks for putting it out there for folks to use!

Here's my question: 
I don't know if there is a way to do this (not documented or I'm stupid) or if there isn't a way to do this, but if I publish a pre-patch/minor/major version, I would like a way for `lerna publish` to be able to figure out those packages still need a non-'pre' publish on subsequent `lerna publish`'s.

Example:
I have a feature branch which has several packages being updated. I want to be able to deploy that branch to an environment and test against it before merging it into my master branch. Therefore, I need to (I think) do a `lerna publish` which will get all those affected components since my build process relies on pulling in the actual npm packages. However, I obviously don't want to do a full version bump since this feature is still in testing. So I use the `pre-X` publish option. This works great!

Then I decide to merge that branch into master after some testing is done. However, now when I run `lerna publish`, it fails to publish anything because in `lerna`'s eyes, nothing has changed. I would still like to do an actual publish though on those packages that have a `pre-X` version number. Of course I can go through figure out which things changed and publish those packages manually, but in a project with a lot of packages, that can be pretty tedious.

Questions:
1. Is there a way to do this currently?
2. If there is, can that be documented in the readme (or if I just missed it, laugh at me and point me towards it)?
3. If there is not a way to do that, is it on a roadmap as a feature, or would you entertain putting that as a possible feature? I think it would be very useful for this kind of workflow.

Thanks again for the great project!

### LINKED ISSUE #956: publish --skip-npm && publish --skip-git never finds updates
Because the tag is written before the call to `publish --skip-git`, it compares the current state against itself, meaning there will never be updates to publish.

### LINKED ISSUE #961: Separate "version" and "publish" commands
Lerna needs a separate `version` command.

### Why
Folks who use npm a lot are probably comfortable with the separation between the `npm version` and `npm publish` commands. For me, this separation is important because it allows me to double- and triple-check things before publishing. When versioning and publishing a monorepo with dozens of packages, the "publish anxiety" factor is even higher. I've worked through some of that anxiety with various permutations of `--skip-git` and `--skip-npm` then `git diff`, but at some point you have to `git reset` and just trust that Lerna is going to do the right thing.

The other problem is that Lerna doesn't make it easy (possible?) to just publish all of the packages using the existing `package.json` versions. AFAICT, it's not possible to skip the increment step, which means that you have to run something like this instead:

```
lerna exec --bail=false npm publish
```

...which will output a lot of scary `npm publish` error messages for any of your packages that haven't been bumped since the last release.

We recently introduced a CI-driven automated publishing workflow in [Primer](/primer/primer-css), and to get around this we resorted to [running](https://github.com/primer/primer-css/blob/master/script/release#L9) a [`try-publish` script](https://github.com/primer/primer-css/blob/master/script/try-publish), which would only run `npm publish` if the version in each module's `package.json` hadn't already been published. Other teams that want to automate publishing will inevitably run into the same issue, and it would stink if they all ended up having to come up with their own solutions for it.

### How
There are a couple of ways to go about this:

1. To maintain backwards compatibility with the current CLI API, the `publish` command could introduce two new options:
    * one that does everything _but_ the git and npm steps, e.g. `lerna publish --skip-publish`
    * one that _only_ publishes without touching any `package.json` files, e.g. `lerna publish --skip-version`

1. If you're comfortable breaking the `publish` CLI for 2.x users, you could avoid adding any additional options to `publish` and just move all of the versioning logic over to a new `version` command.

### LINKED ISSUE #1056: lerna to publish packages with their current version
How do I tell `lerna` to publish packages with **their current version**?
Now `lerna` is always trying to increase the version of packages ...

I need this because I want to divide the publication process into 2 stages:
1. First I want to manually raise the package version only in git in dialog mode:
`lerna publish --skip-npm`
2. Only then with the help of CI I want to publish in npm:
`lerna publish --skip-git`

I do not need to raise the version twice.

## My Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 2.2.0|
| `npm --version`  | 5.4.2|
| `node --version` | v6.11.2|

| OS | Version |
| --- | --- |
| Windows 10 Pro | 1703 |

### LINKED ISSUE #1118: Invalid canary versions for certain SHAs
## Expected Behavior
`--canary` should always generate valid version strings (in accordance with the generally accepted [semver rules](http://semver.org)).

## Current Behavior
`--canary` generates invalid versions if the current SHA's first eight characters are numeric and the first digit is zero. For example, `1.1.0-alpha.21636f26` is valid, but `1.1.0-alpha.08213312` isn't – see rule 9 of the specification ("Numeric identifiers MUST NOT include leading zeroes").

Note that this is not just an issue of just being pedantic – we're using JFrog's Artifactory, and without going into too much detail, it really, *really* doesn't like encountering a version string like that. We're publishing a lot of packages, multiple times a day, and we're hitting this issue quite regularly.

## Possible Solution
Change the template string in `PublishCommand`/`getCanaryVersion()` to be: `${nextVersion}-${preid}${hash}`, or `${nextVersion}-${preid}.commit${hash}`, or something similar.

## Steps to Reproduce
1. Commit until the latest SHA starts with a zero followed by seven digits. This might take a while :)
2. Publish using `--canary`

## Your Environment
| Executable | Version |
| ---: | :--- |
| `lerna --version` | 2.2.0 |
| `npm --version`  | 5.5.1 |
| `yarn --version` | 1.3.2 |
| `node --version` | 8.9.1 |

| OS | Version |
| --- | --- |
| macOS Sierra | 10.12.6 |

### LINKED ISSUE #1385: lerna with CI/CD pipeline
## introduction
We have a monorepo project comprised of 5 packages. We are using lerna to manage package versions, with independent strategy. Overall, apart from one particular big surprise, the experience with lerna is good. 

## our  pipeline setup
5 developers work on different features, once their work is finished, code is merged into master, which will trigger a build in Bamboo. The bamboo build server will build the code, bump the version of changed packages and publish them into a private npm registry.

## the problem
the problem happens when people merge their code into master while there is an active build in progress.
For example, lets say my project has `A@1.0.0`, `B@1.0.0` and `C@1.0.0` to begin with.  Here is what happened (in time order)
1. developer A merged his code into master which contains `chore(A)` and `feat(B)`. Bamboo picked it up and starts build#1;
2. while build#1 is under way, developer B merged his code which contains `feat(C)` and `chore(A)`. This triggers another build#2.  Now we have both build#1 and build#2 running at the same time;
3. build#1 bumped packages to `A@1.0.1` and `B@1.1.0`; and successfully published them into npm registry. But when it tries to `git push` all those  `package.json` files and `changelog.md`s back, it failed because it doesn't contains latest change(due to developer B 's merge). So build#1 failed with  `A@1.0.1` and `B@1.1.0`  published into npm registry. No commits or tags were pushed to origin.
4. build#2 bumped packages to `C@1.1.0` , `A@1.0.1` and `B@1.1.0` (note here that `B` will be picked up as well because build#1 didn't push the new tags. even if it did, build#2 may not be able to see it due to timing condition). It successfully published `C@1.1.0` but failed on publishing `A@1.0.1` because that package already exists. So build#2 also failed with `C@1.1.0` published. No commits or tags were pushed to origin. 

Now, imagine you sit in front a dev machine and do a `git pull`. All your local `package.json` files are still stating `A@1.0.0`, `B@1.0.0` and `C@1.0.0`, no new git tags introduced,  but all 3 npm packages have been published. What do you do?  😵 😵 😱  😱 

## possible solutions : disable concurrent builds in bamboo
you may think that by making bambo build sequential the problem may be solved. But it wont' help because build#1 still fails due to not having developer B 's merge. build#2 also will fail because build#1 didn't push its local commits&tags.  so disabling concurrent build literally has no effect at all. 

## your solution
I do have a solution but it's not ideal and I would like to hear from others before I share it here.

I'm pretty sure i'm not alone, as it seems to be the standard use case of lerna. What is the official way of using lerna in a CI/CD pipeline??

### LINKED ISSUE #1483: Ignore `v` in `--repo-version`
I want to publish with `circleci` and `git tag`. Normally I use `git tag` like `v1.0.0`, and publish in `circleci` like `yarn lerna publish --repo-version ${CIRCLE_TAG}`. As a result, `yarn lerna publish` will use `v1.0.0` to publish packages, but `--new-version` of `yarn publish` can not be a version with `v`. Maybe `v` should be ignored with `--repo-version`?

## Expected Behavior

```sh
yarn publish --repo-version v1.0.0
// publish packages with version `1.0.0`.
```

## Current Behavior

```sh
yarn publish --repo-version v1.0.0
// publish packages with version `v1.0.0`.
```

## Possible Solution

Add `repoVersion.replace(/^v/, '')` in [here](https://github.com/lerna/lerna/blob/master/commands/publish/index.js#L305).

## Steps to Reproduce (for bugs)

```sh
yarn lerna publish --repo-version v1.0.0
```

<details><summary>lerna.json</summary><p>

```json
<!-- Please paste your `lerna.json` here -->
{
  version: 1.0.0,
  commands: { 
    publish: { 
      ignore: [ 
        *.md,
        __tests__/**
      ]
    }
  },
  lerna: 2.11.0,
  npmClient: yarn,
  npmClientArgs: [ 
    --no-lockfile
  ],
  packages: [ 
    packages/*
  ]
}
```
</p></details>

<details><summary>lerna-debug.log</summary><p>

```txt
<!-- If you have a `lerna-debug.log` available, please paste it here -->
<!-- Otherwise, feel free to delete this <details> block -->
```
</p></details>

## Context

I hope this can publish in `circleci` with my `git tag`, and I want to still have `v` in my `git tag`. Like `standard-version`, this will ignore `v` when I use `--release-as`. Maybe `lerna` can ignore, too.

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 2.11.0 |
| `npm --version`  | 6.1.0 |
| `yarn --version` | 1.7.0 |
| `node --version` | 10.2.1 |

| OS | Version |
| --- | --- |
| macOS High Sierra | 10.13.5 |

### PR DESCRIPTION
**Pull request by evocateur**
## Description

To facilitate further refactoring, the gigantic PublishCommand needs to be split in two. Several flags were deprecated and replaced with clearer options, but remarkably only two _technically_ breaking changes resulted from ~2 weeks of work.

BREAKING CHANGE:
* `--preid` now defaults to "alpha" during prereleases:

  The previous default for this option was undefined, which led to an awkward "1.0.1-0" result when passed to `semver.inc()`.

  The new default "alpha" yields a much more useful "1.0.1-alpha.0" result. Any previous prerelease ID will be preserved, just as it was before.

* `--no-verify` is no longer passed to `git commit` by default, but controlled by the new `--commit-hooks` option:

  The previous behavior was too overzealous, and the new option operates exactly like the corresponding [npm version](https://docs.npmjs.com/cli/version#commit-hooks) option of the same name.

  As long as your pre-commit hooks are properly scoped to ignore changes in package.json files, this change should not affect you. If that is not the case, you may pass `--no-commit-hooks` to restore the previous behavior.

## Motivation and Context

Fixes #277
Fixes #936
Fixes #956
Fixes #961
Fixes #1056
Fixes #1118
Fixes #1385
Fixes #1483
Fixes #1494

This PR lays the groundwork for #1091 (and #1513) by making `lerna publish` only _directly_ worrying about publish itself, not all those other complicated versioning decisions.

## How Has This Been Tested?

Ported unit tests, with "full-bleed" integration tests of `npm publish` thanks to the new `--dry-run` flag in npm@6.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (internal tools)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by morgs32** [comment]
@evocateur I salute you this looks so promising. I am curious about the bit of documentation that strongly discourages running `lerna version` from a branch. I kind of figured that I'd manually be incrementing version on a branch and in a PR, and then on merge to `master` I'd `lerna publish from-git`. But is that not a wise idea? What sort of process would you suggest or what did you have in mind?

**Comment by evocateur** [comment]
@morgs32 My biggest concern with making tagged releases in feature branches is that the tags get mangled when merging.

I'm not sure if rebase merge would be any safer, since if it was rebasing onto upstream commits, the SHAs would change and thus the annotated tag(s) would be "broken", as the SHA it points to no longer exists. However, I haven't tried rebase merge in a long time, and I don't recall if `git rebase` itself preserves tags.

One of my goals in the near future is to improve the "auto-publish from CI" experience, perhaps with a subcommand `lerna publish release`. I envision some configurable branch awareness that would publish untagged prereleases on feature branches (or PRs only?) and then on `master` (or `latest`, whatever your "default" branch is configured to be) it would publish with tags, even creating github releases with changelogs (if you're using conventional commits).

So please let me know if `lerna publish from-git` works after a rebase merge, I would like to figure this out, even if it is "just" a clarification to the documentation!

File: commands/version/README.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+### `--allow-branch <glob>`
+
+A whitelist of globs that match git branches where `lerna version` is enabled.
+It is easiest (and recommended) to configure in `lerna.json`, but it is possible to pass as a CLI option as well.
+
+```json
+{
+  "command": {
+    "publish": {
```
Comment (danielo515) [inline commands/version/README.md]: Is this correct @evocateur ? Do you configure `lerna version` by configuring publish on `lerna.json` ? Seems confusing

**Comment by evocateur** [comment]
It's for backward-compatibility. The `lerna version` docs probably should be updated to use the correct namespace, though.

**Pull request merged**